tfjs-converter/src/executor/graph_executor.ts

```diff
@@ -318,44 +330,74 @@ export class GraphExecutor implements FunctionExecutor {
       return;
     }
 
-    tensorMap[nodeName].forEach(tensor => {
-      if (tensor != null) {
-        intermediateTensorConsumerCount[tensor.id] =
-            (intermediateTensorConsumerCount[tensor.id] || 0) +
-            node.children.length;
+    for (const tensor of tensorMap[nodeName]) {
+      if (tensor == null) {
+        continue;
       }
-    });
-    node.inputs.forEach(input => {
+      intermediateTensorConsumerCount[tensor.id] =
+          (intermediateTensorConsumerCount[tensor.id] || 0) +
+          node.children.length;
+    }
+
+    for (const input of node.inputs) {
       // Skip any control flow nodes, since its dependency is tricky to track
       // correctly.
-      if (input.category !== 'control') {
-        const tensors =
-            getTensorsForCurrentContenxt(input.name, tensorMap, context);
-        if (tensors != null) {
-          tensors.forEach(tensor => {
-            if (tensor && !tensor.kept && !tensorsToKeep.has(tensor.id)) {
-              const count = intermediateTensorConsumerCount[tensor.id];
-              if (count === 1) {
-                tensor.dispose();
-                delete intermediateTensorConsumerCount[tensor.id];
-              } else if (count != null) {
-                // only intermediate nodes has count set, inputs and weights
-                // are not.
-                intermediateTensorConsumerCount[tensor.id]--;
-              }
-            }
-          });
+      if (isControlFlow(input)) {
+        continue;
+      }
+
+      const tensors =
+          getTensorsForCurrentContext(input.name, tensorMap, context);
+      if (tensors == null) {
+        continue;
+      }
+
+      for (const tensor of tensors) {
+        if (!tensor || tensor.kept || tensorsToKeep.has(tensor.id)) {
+          continue;
+        }
+
+        const count = intermediateTensorConsumerCount[tensor.id];
+        if (count === 1) {
+          tensor.dispose();
+          delete intermediateTensorConsumerCount[tensor.id];
+        } else if (count != null) {
+          // only intermediate nodes has count set, inputs and weights
+          // are not.
+          intermediateTensorConsumerCount[tensor.id]--;
         }
       }
-    });
+    }
+  }
+
+  private checkTensorForDisposalWithNodeLiveUntilInfo(
+      node: Node, tensorMap: NamedTensorsMap, tensorsToKeep: Set<number>,
+      outputNodeNameSet: Set<string>, liveUntilNodes?: Node[]) {
+    // Skip output nodes and any control flow nodes, since its dependency is
+    // tricky to track correctly.
+    if (isControlFlow(node) || outputNodeNameSet.has(node.name)) {
+      return;
+    }
+    if (liveUntilNodes == null) {
+      return;
+    }
+
+    for (const node of liveUntilNodes) {
+      for (const tensor of tensorMap[node.name]) {
+        if (!tensor || tensor.kept || tensorsToKeep.has(tensor.id)) {
+          continue;
+        }
+        tensor.dispose();
+      }
+    }
```

Comment:
I intentionally make this function keep similar signature and name as the original `checkNodesForDisposal`, to maintain the parity of tensor disposing functions in sync and async graph executor. I want to make it easier for future maintainer to understand that they are the same, at least from the aspect of intentions and behaviors.

I think a better rewrite may be renaming both of them to something like "function disposeNodes...(...)", since the original function not only "checks" the tensors but also disposes tensors in it. What do you think?

For factoring out some stuff, I don't see huge gains on making a separate function for that single if check. I'm thinking to rewrite the async disposing algorithm to utilize the node dependency and reduce loops over tensor edges, but I haven't figured out a better algo (and I need to learn more about async control flow to make sure the new algo won't break it). Maybe we can have a better picture of which parts can be shared then.